pipeline-runner/R/gem2s-2-load_user_files.R

```diff
@@ -234,3 +254,84 @@ format_annot <- function(annot_list) {
+#  0 is SYMBOL/ SYMBOL
+#  1 is IDS/SYMBOL
+#  2 is IDS/IDS
+equalize_annotation_types <- function(annot_list, counts_list, features_types_list, samples) {
+  if (any(features_types_list == 2) && any(features_types_list == 0) && !any(features_types_list == 1)) stop("Incompatible features detected.")
+
+  if (any(features_types_list == 1) && (any(features_types_list == 2) || any(features_types_list == 0))) {
+    annots_with_ids <- unique(do.call("rbind", annot_list[features_types_list == 1]))
+
+    annots_with_ids <- annots_with_ids[!duplicated(annots_with_ids$input), ]
+
+    for (sample in samples) {
+      if (features_types_list[[sample]] == 0 || features_types_list[[sample]] == 2) {
+        sample_annot <- annot_list[[sample]]
+
+        # Try to replace input column (currently symbols) in sample_annot with ids from annots_with_ids
+        if (features_types_list[[sample]] == 0) {
+
+          matched_symbols_index <- match(sample_annot$input,annots_with_ids$name)
+          is_in_annot_list <- which(sample_annot$input %in% annots_with_ids$name)
+
+          sample_annot$input[is_in_annot_list] <- annots_with_ids$input[na.omit(matched_symbols_index)]
+
+          #This avoids duplicates after combining with the annotated df.
+          #Leads to a mismatch in genes between samples but it seems like the best solution
+          sample_annot$input <- make.unique(sample_annot$input)
+
+          #The counts have been loaded with gene symbols so we need to replace the rownames with the ids
+          rownames(counts_list[[sample]]) <- sample_annot$input
+        }
+
+        # Try to replace names column (currently ids) in sample_annot with symbols from annots_with_ids
+        if (features_types_list[[sample]] == 2) {
+          matched_ids_index <- match(sample_annot$input,annots_with_ids$input)
+          is_in_annot_list <- which(sample_annot$input %in% annots_with_ids$input)
+
+          sample_annot$name[is_in_annot_list] <- annots_with_ids$name[na.omit(matched_ids_index)]
+        }
+
+        annot_list[[sample]] <- sample_annot
+      }
+    }
+  }
+  return(list(counts_list=counts_list, annot_list=annot_list))
+}
+
+#' extract_features_types
+#'
+#' Determines the type of an annot data frame.
+#' Classifies the columns into either ensemblIds or symbols, and extracts a number that represents the combination.
+#'
+#' @param annot
+#'
+#' @return
+#' -1 is SYMBOl/IDS
+#' 0 is SYMBOL/ SYMBOL
+#' 1 is IDS/SYMBOL
+#' 2 is IDS/IDS
+#'
+#'
+#' @examples
+extract_features_types <- function(annot) {
+  features_types <- list()
+
+  annot_c1 <- annot[, 1]
+  is_ens <- annot_c1[substr(annot_c1, 1, 3) == "ENS"]
+  features_types[[1]] <- length(is_ens) >= length(annot_c1) - length(is_ens)
+
+  annot_c2 <- annot[, 2]
+  is_ens <- annot_c2[substr(annot_c2, 1, 3) == "ENS"]
+  features_types[[2]] <- length(is_ens) >= length(annot_c2) - length(is_ens)
+
+  if (features_types[[1]] == FALSE && features_types[[2]] == TRUE) {
```

Comment:
I don't get this `feature_types` have been one of those SYM_SYM, IDS_IDS, etc... values but here it's suddenly a boolean? it looks like it's a boolean resulting from comparing `lenghts` so naming it `types` does not make sense.

Reply:
I made this to make this function easier to write instead of ending in three:

if(a) return a
if(b) return b
if(c) return c

Reply:
Maybe that option will be clearer though

Reply:
But basically in this function we are still working with booleans and from the combination of bools we return an int that represents the current state of the features file

Reply:
I do love the idea of replacing all numbers for constant numbers as in 

IDS_IDS
SYM_SYM

Etc.

Reply:
I was going to add a comment to change it to a dictionary but constants sound much better

Reply:
why the need for ints? Just make the constants strings. That way when we debug and print `feature_types` we get `IDS_IDS` instead of `0` or `-1`.

Reply:
replaced all ints with character variables defined in sysdata.R